modules/rostests/kmtests/hal/HalPortIo.c

```diff
@@ -0,0 +1,141 @@
+/*
+ * PROJECT:     ReactOS Kernel-Mode Tests
+ * LICENSE:     LGPL-2.1-or-later (https://spdx.org/licenses/LGPL-2.1-or-later)
+ * PURPOSE:     Tests for string I/O intrinsic functions
+ * COPYRIGHT:   Copyright 2025 Dmitry Borisov <di.sean@protonmail.com>
+ */
+
+/* INCLUDES *******************************************************************/
+
+#include <kmt_test.h>
+
+#define NDEBUG
+#include <debug.h>
+
+/* GLOBALS ********************************************************************/
+
+/* Compile with the highest optimization level to trigger a bug (See CORE-20078) */
+#ifdef _MSC_VER
+#pragma optimize("gst", on)
+#pragma auto_inline(on)
+#else
+#pragma GCC optimize("O3")
+#endif
+
+/* Used to isolate the effects of intrinsics from each other */
+typedef struct _TEST_CONTEXT
+{
+    PVOID OldBuffer;
+    UCHAR Pad[78];
+    ULONG Port;
+    ULONG Size;
+    PVOID Buffer;
+} TEST_CONTEXT, *PTEST_CONTEXT;
+
+/* FUNCTIONS ******************************************************************/
+
+static
+DECLSPEC_NOINLINE // Isolate the effects of intrinsics from each other
+VOID
+TestWriteStringUshort(
+    _In_ PTEST_CONTEXT Context)
+{
+    __outwordstring(Context->Port, Context->Buffer, Context->Size / sizeof(USHORT));
+
+    /*
+     * The 'rep outsw' instruction increments or decrements the address in ESI by Size * 2.
+     * Test for CORE-20078: the ESI value should be preserved across calls.
```

Comment:
Some notes:
* I preferred my comment here, and yours would better apply to `__outwordstring()` line itself.
* This behavior is common to all 6 opcodes. It's "unexpected" the bug affects/shows_up only 1 case.
Well, we'll probably want to fix all functions, to be safe...
https://www.felixcloutier.com/x86/ins:insb:insw:insd
https://www.felixcloutier.com/x86/outs:outsb:outsw:outsd
* Then, "instead" of (optimized) `ok_eq_pointer()`, should you explicitly check DI/SI register?

Reply:
> This behavior is common to all 6 opcodes. It's "unexpected" the bug affects/shows_up only 1 case.

This applies to all `rep outs*` operations as well.

> should you explicitly check DI/SI register?

```
void __outwordstring(
   unsigned short Port,
   unsigned short* Buffer,
   unsigned long Count
);
```
I'm not sure if this would be possible, since `Port` and `Count` are passed by value.